src/DataRenderer.tsx

```diff
@@ -85,33 +89,67 @@ function ExpandableObject({
   const lastIndex = data.length - 1;
 
   const onKeyDown = (e: React.KeyboardEvent<HTMLSpanElement>) => {
-    if (e.key === ' ') {
+    if (e.key === 'ArrowRight' || e.key === 'ArrowLeft') {
+      e.preventDefault();
+      setExpanded(e.key === 'ArrowRight');
+    } else if (e.key === 'ArrowUp' || e.key === 'ArrowDown') {
       e.preventDefault();
-      toggleExpanded();
+      const direction = e.key === 'ArrowUp' ? -1 : 1;
+
+      if (!outerRef.current) return;
+      const buttonElements = Array.from(
+        outerRef.current.querySelectorAll<HTMLElement>('[role=button]')
+      );
+      const currentIndex = buttonElements.findIndex((el) => el.tabIndex === 0);
+      if (currentIndex < 0) return;
+
+      const nextIndex = (currentIndex + direction + buttonElements.length) % buttonElements.length; // auto-wrap
+      buttonElements[currentIndex].tabIndex = -1;
+      buttonElements[nextIndex].tabIndex = 0;
+      buttonElements[nextIndex].focus();
     }
   };
 
+  const onClick = () => {
+    toggleExpanded();
+
+    const buttonElement = expanderButtonRef.current;
```

Comment:
Thanks for a lot of efforts!
Just couple more questions :) Do you think if we will update tabIndex and focus separately (for instance using setTimeout) it will help to render the changes faster and than change only the tabIndex. Or it will cause double work for the React virtual DOM and in the end it will be more slow overall?

```js
setTimeout(() => {
  const prevButtonElement = outerRef.current?.querySelector<HTMLElement>(
    '[role=button][tabindex="0"]'
  );
  if (prevButtonElement) {
    prevButtonElement.tabIndex = -1;
  }
  buttonElement.tabIndex = 0;
  buttonElement.focus();
}, 0);
```

Reply:
I'm not sure what you mean exactly here?

To summarize, this `tabIndex`/`focus` code in `onClick` is to handle the situation where you're currently tab-focused on a nested expand button, but then you click a parent node button to collapse. Now that inner node that had `tabIndex=0` has disappeared, so the `tabIndex=0` needs to be redirected to somewhere. And then the `focus` on the expand button makes it so you can proceed to use the arrow keys even if it was the text that was clicked to collapse.

If you have a different idea for how to accomplish that, I'm all ears!

Reply:
I mean if we can split this logic into 2 independent (async) actions, one which immediately visible for the user (`toggleExpanded();`) and second (`tabIndex`/`focus`) which is less noticable. So expanding node will feel slightly more responsible. I was worrying how React will handle (will it start update DOM only after we handle tabIndex/focus anyway?) this but seems like tabIndex/focus logic uses only pure JS which is not related to the virtual DOM.

For now for 300Kb JSON data adding tabIndex/focus logic is hardly noticable so I think we can merge as it is and try to run some perf tests.